File: packaging/windows/launcher.py
"""Hermes WebUI — Windows launcher.

Compiled to ``Hermes.exe`` with PyInstaller. It lives at the install root next
to ``python\\``, ``webui\\`` and ``agent\\``. Responsibilities:

  1. Point the WebUI at the bundled embedded Python and agent source.
  2. Start ``webui\\server.py`` with the embedded interpreter (no console).
  3. Wait until the HTTP server answers, then open the default browser.
  4. On exit, tear down the server process tree (psutil) and best-effort stop
     any background Hermes gateway the agent may have spawned.

This is the Windows counterpart of packaging/macos/supervisor.py + the Swift
shell, collapsed into one small launcher.
"""
from __future__ import annotations

import os
import sys
import time
import subprocess
import urllib.request
import webbrowser
from pathlib import Path


def _install_root() -> Path:
    # When frozen by PyInstaller, sys.executable is <install>\Hermes.exe.
    if getattr(sys, "frozen", False):
        return Path(sys.executable).resolve().parent
    return Path(__file__).resolve().parent


ROOT = _install_root()
PYTHON = ROOT / "python" / "python.exe"
WEBUI = ROOT / "webui"
AGENT = ROOT / "agent"
SERVER = WEBUI / "server.py"

HOST = os.environ.get("HERMES_WEBUI_HOST", "127.0.0.1")
PORT = int(os.environ.get("HERMES_WEBUI_PORT", "8787"))


def _local_app_data() -> Path:
    base = os.environ.get("LOCALAPPDATA") or str(Path.home() / "AppData" / "Local")
    return Path(base) / "hermes"


def _build_env() -> dict:
    env = dict(os.environ)
    env["HERMES_HOME"] = env.get("HERMES_HOME") or str(_local_app_data())
    # config.py (_discover_agent_dir) prefers HERMES_WEBUI_AGENT_DIR and injects
    # it onto the FRONT of sys.path, so the WebUI can `import gateway`/run_agent
    # in-process from the bundled agent source.
    env.setdefault("HERMES_WEBUI_AGENT_DIR", str(AGENT))
    env.setdefault("HERMES_WEBUI_HOST", HOST)
    env.setdefault("HERMES_WEBUI_PORT", str(PORT))
    # No separate agent venv is created — every dependency (WebUI + agent) lives
    # in this embedded interpreter. Tell the WebUI to reuse it for any helper it
    # launches, instead of hunting for an agent venv that does not exist.
    env["HERMES_WEBUI_PYTHON"] = str(PYTHON)
    return env


def _wait_until_up(timeout: float = 90.0) -> bool:
    url = f"http://{HOST}:{PORT}/"
    deadline = time.time() + timeout
    while time.time() < deadline:
        try:
            with urllib.request.urlopen(url, timeout=2) as r:
                if r.status < 500:
                    return True
        except Exception:
            time.sleep(0.5)
    return False


def _terminate_tree(proc: subprocess.Popen) -> None:
    try:
        import psutil  # bundled into the launcher exe
    except Exception:
        psutil = None
    if psutil is not None:
        try:
            parent = psutil.Process(proc.pid)
            procs = parent.children(recursive=True) + [parent]
            for p in procs:
                try:
                    p.terminate()
                except Exception:
                    pass
            _gone, alive = psutil.wait_procs(procs, timeout=8)
            for p in alive:
                try:
                    p.kill()
                except Exception:
                    pass
            return
        except Exception:
            pass
    try:
        proc.terminate()
    except Exception:
        pass


def _stop_gateway(env: dict) -> None:
    # Best-effort: if the agent started a background gateway, stop it so nothing
    # lingers after the window closes. Failure here is harmless.
    run_agent = AGENT / "run_agent.py"
    if not run_agent.exists():
        return
    try:
        subprocess.run(
            [str(PYTHON), str(run_agent), "gateway", "stop"],
            cwd=str(AGENT), env=env, timeout=30,
            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL,
        )
    except Exception:
        pass


def main() -> int:
    if not PYTHON.exists():
        sys.stderr.write(f"Embedded Python not found: {PYTHON}\n")
        return 2
    if not SERVER.exists():
        sys.stderr.write(f"WebUI server not found: {SERVER}\n")
        return 2

    env = _build_env()
    log_dir = Path(env["HERMES_HOME"])
    log_dir.mkdir(parents=True, exist_ok=True)
    log_path = log_dir / "webui.log"

    creationflags = getattr(subprocess, "CREATE_NO_WINDOW", 0) if os.name == "nt" else 0
    with open(log_path, "ab") as logf:
        proc = subprocess.Popen(
            [str(PYTHON), str(SERVER)],
            cwd=str(WEBUI), env=env,
            stdout=logf, stderr=logf,
            creationflags=creationflags,
        )

    try:
        if _wait_until_up():
            webbrowser.open(f"http://{HOST}:{PORT}/")
        else:
            sys.stderr.write("WebUI did not become healthy in time; see webui.log\n")
        proc.wait()
    except KeyboardInterrupt:
        pass
    finally:
        if proc.poll() is None:
            _terminate_tree(proc)
        _stop_gateway(env)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: packaging/windows/launcher.c
/* Hermes WebUI — Windows console launcher (portable/green build).
 *
 * Cross-compiled to Hermes.exe from macOS/Linux with:
 *   zig cc -target x86_64-windows-gnu launcher.c -o Hermes.exe -lshell32
 *
 * Lives at the portable root next to python\, webui\, agent\. It:
 *   1. Points the WebUI at the bundled embedded Python and agent source.
 *   2. Starts python\python.exe webui\server.py inside a Job object.
 *   3. Opens the default browser at the WebUI URL.
 *   4. Blocks until the server exits; closing this console window terminates
 *      the whole process tree (JOB_OBJECT_LIMIT_KILL_ON_JOB_CLOSE), so nothing
 *      lingers.
 */
#include <windows.h>
#include <shellapi.h>
#include <stdio.h>
#include <string.h>

int main(void) {
    char exePath[MAX_PATH];
    GetModuleFileNameA(NULL, exePath, MAX_PATH);

    char dir[MAX_PATH];
    snprintf(dir, sizeof(dir), "%s", exePath);
    char *slash = strrchr(dir, '\\');
    if (slash) *slash = '\0';

    char python[MAX_PATH * 2], webui[MAX_PATH * 2], server[MAX_PATH * 2], agent[MAX_PATH * 2];
    snprintf(python, sizeof(python), "%s\\python\\python.exe", dir);
    snprintf(webui,  sizeof(webui),  "%s\\webui", dir);
    snprintf(server, sizeof(server), "%s\\webui\\server.py", dir);
    snprintf(agent,  sizeof(agent),  "%s\\agent", dir);

    /* config.py (_discover_agent_dir) prefers HERMES_WEBUI_AGENT_DIR and injects
     * it onto the front of sys.path so the WebUI can import the agent in-process.
     * No separate agent venv exists — all deps live in the embedded interpreter. */
    SetEnvironmentVariableA("HERMES_WEBUI_AGENT_DIR", agent);
    SetEnvironmentVariableA("HERMES_WEBUI_PYTHON", python);

    HANDLE job = CreateJobObjectA(NULL, NULL);
    if (job) {
        JOBOBJECT_EXTENDED_LIMIT_INFORMATION jeli;
        memset(&jeli, 0, sizeof(jeli));
        jeli.BasicLimitInformation.LimitFlags = JOB_OBJECT_LIMIT_KILL_ON_JOB_CLOSE;
        SetInformationJobObject(job, JobObjectExtendedLimitInformation, &jeli, sizeof(jeli));
    }

    char cmdline[MAX_PATH * 4];
    snprintf(cmdline, sizeof(cmdline), "\"%s\" \"%s\"", python, server);

    STARTUPINFOA si;
    PROCESS_INFORMATION pi;
    memset(&si, 0, sizeof(si));
    si.cb = sizeof(si);
    memset(&pi, 0, sizeof(pi));

    printf("Starting Hermes WebUI...\n");
    fflush(stdout);

    if (!CreateProcessA(NULL, cmdline, NULL, NULL, FALSE,
                        CREATE_SUSPENDED, NULL, webui, &si, &pi)) {
        printf("ERROR: failed to start server (error %lu)\n", (unsigned long)GetLastError());
        printf("Looked for: %s\n", python);
        printf("Press Enter to exit.\n");
        getchar();
        return 1;
    }
    if (job) AssignProcessToJobObject(job, pi.hProcess);
    ResumeThread(pi.hThread);

    /* Honor HERMES_WEBUI_PORT (default 8787), same as the WebUI/server reads it,
     * so the launched URL never drifts from the actual bind port. */
    char port[16] = "8787";
    GetEnvironmentVariableA("HERMES_WEBUI_PORT", port, sizeof(port));
    char url[64];
    snprintf(url, sizeof(url), "http://127.0.0.1:%s/", port);

    Sleep(4000);
    ShellExecuteA(NULL, "open", url, NULL, NULL, SW_SHOWNORMAL);
    printf("Hermes is running at %s\n", url);
    printf("Close this window to stop Hermes.\n");
    fflush(stdout);

    WaitForSingleObject(pi.hProcess, INFINITE);
    if (job) CloseHandle(job);
    CloseHandle(pi.hThread);
    CloseHandle(pi.hProcess);
    return 0;
}
